libpromises/evalfunction.c

```diff
@@ -6390,110 +6461,23 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
         {
             requested_mode = "CSV";
         }
-        else
-        {
-            requested_mode = "JSON";
-        }
-    }
-
-    if (strcmp("CSV", requested_mode) == 0)
-    {
-        size_t size_max = 50 * (1024 * 1024);
-
-        size_t byte_count = 0;
-
-        FILE *fin = safe_fopen(input_path, "r");
-        if (fin == NULL)
-        {
-            Log(LOG_LEVEL_VERBOSE, "%s cannot open the CSV file '%s' (fopen: %s)",
-                fp->name, input_path, GetErrorStr());
-            return FnFailure();
-        }
-
-        JsonElement *json = JsonArrayCreate(50);
-        int linenumber = 0;
-        char *line;
-        while ((line = GetCsvLineNext(fin)) != NULL)
+        else if (0 == strcmp(fp->name, "readenvfile"))
         {
-            ++linenumber;
-
-            byte_count += strlen(line);
-            if (byte_count > size_max)
-            {
-                Log(LOG_LEVEL_VERBOSE, "%s: line %d from CSV file '%s' exceeded byte limit %lu, done with file",
-                    fp->name, linenumber, input_path, (long unsigned int)size_max);
-                free(line);
-                break;
-            }
-
-            Seq *list = SeqParseCsvString(line);
-            free(line);
-
-            if (list != NULL)
-            {
-                JsonElement *line_arr = JsonArrayCreate(SeqLength(list));
-
-                for (size_t i = 0; i < SeqLength(list); i++)
-                {
-                    JsonArrayAppendString(line_arr, SeqAt(list, i));
-                }
-
-                SeqDestroy(list);
-                JsonArrayAppendArray(json, line_arr);
-            }
+            requested_mode = "ENV";
         }
-
-        bool atend = feof(fin);
-        fclose(fin);
-
-        if (!atend)
+        else
```

Comment:
Thanks, it's now more clear; the complexity isn't trivial since `FnCallReadData()` is called from policy functions `readdata()`, `readcsv()`, `readjson()`, `readyaml()`, `readenvfile()`. Mode of operation is modified according to function name and `requested_mode` parameter.